### PR TITLE
feat: shipping discount

### LINKED ISSUE #6767: As an operator I want Shipping Discount promotions to stack unless they are actioned on the same Shipping Method
By default an operator will want promos of the Shipping promotion type to stack with each other unless the stacking promos have their action set for the same Shipping Method

### LINKED ISSUE #6773: Promotions: Add shipping discounts
This needs to be added as a new Promotion Type "Shipping Discount"

Add the ability to add a discount (either as a flat amount of as a percentage) to a particular shipping method.

Extra parameters:
* I should be able to include/exclude products from this discount
* I should be able to discount a particular shipping method e.g. UPS Blue, but not UPS Ground
* I should be able to specify the method match parameter optionally as regex so that minor changes don't break the match

The work to complete this ticket should also include any changes to the API to make this work.

### PR DESCRIPTION
**Pull request by vanpho93**
Resolves #6773 #6767
Impact: **major**
Type: **feature**

Adding shipping discount with an option to prevent using 2 shipping discounts for one shipment.

Loom: https://www.loom.com/share/0ea5076ca3004ae9b1be6b3880a32e86

### CODE REVIEW

File: packages/api-plugin-promotions-discounts/src/discountTypes/shipping/applyShippingDiscountToCart.js
Comment (brent-hoover) [inline packages/api-plugin-promotions-discounts/src/discountTypes/shipping/applyShippingDiscountToCart.js:39]: Are we applying the shipping discount to items? Not sure that's how we want to do that

File: packages/api-plugin-promotions-discounts/src/discountTypes/shipping/applyShippingDiscountToCart.js
Comment (brent-hoover) [inline packages/api-plugin-promotions-discounts/src/discountTypes/shipping/applyShippingDiscountToCart.js]: This must be available in a util lib somewhere rather than copy-pasting it here?

**Review by brent-hoover — COMMENTED** [review]
Seems ok just have some questions:

1. We are taking the discount on the shipping charge, not items right?
2. We are taking the discount only on shipping and not on handling, correct?
3. Have we taken into account when we may have a split shipment (i.e. two items in the shipping array)?

File: packages/api-plugin-promotions-discounts/src/discountTypes/shipping/applyShippingDiscountToCart.js
Comment (vanpho93) [reply packages/api-plugin-promotions-discounts/src/discountTypes/shipping/applyShippingDiscountToCart.js:39]: No, we are applying it to shipping only, as it shows on the Loom demo.

**Comment by vanpho93** [comment]
> Seems ok just have some questions:
> 
> 1. We are taking the discount on the shipping charge, not items right?
> 2. We are taking the discount only on shipping and not on handling, correct?
> 3. Have we taken into account when we may have a split shipment (i.e. two items in the shipping array)?

1. => Yes, on the shipping charge
2. => Do we want to exclude handling? Currently handling get discounted too but I can exclude it
3. => We have it but we aren't able to demo it with the API because we didn't support split shipment in API level

**Comment by brent-hoover** [comment]
@vanpho93 Ok, double-checking whether we should discount handling or not, so hold on this.

**Comment by brent-hoover** [comment]
@vanpho93 Clarified with @brendan-shea-pm , discounts should only be taken against shipping, not handling

**Review by brent-hoover — CHANGES_REQUESTED** [review]
Discounts don't display in quotes which is critical

**Comment by brent-hoover** [comment]
@vanpho93 Can we add a shipping discount promotion to the sample data as well?

**Review by brent-hoover — CHANGES_REQUESTED** [review]
Shipping quotes are not updated when a promotion becomes active after the cart already has quotes on it. Brian has identified the problem already and is working on a fix

File: packages/api-plugin-promotions/src/startup.js
Comment (brent-hoover) [inline packages/api-plugin-promotions/src/startup.js]: It should not be one but since we're going to make it an env-var it can stay.

**Review by brent-hoover — APPROVED** [review]
Great work

**Comment by brent-hoover** [comment]
@vanpho93 In the future please wait for all reviewers I've assigned to approve before merging

**Pull request merged**
